### PR TITLE
Improve the way OIDC init failures are handled

### LINKED ISSUE #16725: Oidc doesn't recover when auth-server was not avaliable at startup
## Describe the bug

When runing quarkus with oidc enabled, the application starts even when the oidc-server is currently not avaliable. Each request to the api results in a error:

`io.quarkus.oidc.OIDCException: Tenant configuration has not been resolved
	at io.quarkus.oidc.runtime.OidcAuthenticationMechanism.resolve(OidcAuthenticationMechanism.java:61)
	at io.quarkus.oidc.runtime.OidcAuthenticationMechanism.authenticate(OidcAuthenticationMechanism.java:40)
	at io.quarkus.oidc.runtime.OidcAuthenticationMechanism_Subclass.authenticate$$superaccessor2(OidcAuthenticationMechanism_Subclass.zig:512)
	at io.quarkus.oidc.runtime.OidcAuthenticationMechanism_Subclass$$function$$2.apply(OidcAuthenticationMechanism_Subclass$$function$$2.zig:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:63)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:51)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(InvocationInterceptor_Bean.zig:521)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	...`

Even when the oidc-server gets avaliable after some time, the requests aren't recovering from this error. Not even after the time defined in `quarkus.oidc.token.forced-jwk-refresh-interval`.

### Expected behavior

The requests should either recover after the oidc-server is back or the application should not start if the oidc-server is not avaliable.

### Configuration
```properties
# Authentication
quarkus.oidc.auth-server-url=https://the-server/auth/realms/local-dev
quarkus.oidc.client-id=the-client
```

### Quarkus version or git rev
1.13.2.Final

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #16725

This PR:

- restores the OIDC server failing to start if `auth-server-url` is invalid and logs the message
- keeps attempting recovering the connection if the URL is valid when the requests arrive and logs the message
- if `auth-server-url` is empty and it is a DEV mode - lets server to start - this is typical scenario for starting with dev mode after for ex generating a test project - but fails to start if it is not a DEV mode
- test showing the recovery in action is added - the mock server is started after the Quarkus endpoint has been initialized
- updated the retry code to report `ConnectException` as opposed to `IllegalStateException` which is reported right now when the retries are exhausted

 Going forward we can also do:
- always fail as it used to be a case - but only if the system property is set
- let users handle the OIDC startup events - and fail only of the user handler confirms it is needed, etc

### CODE REVIEW

**Comment by sberyozkin** [comment]
There was a Keycloak image pull problem which cause a test failure - but I had to restart

**Comment by sberyozkin** [comment]
Seeing 
```
Failed to execute goal io.fabric8:docker-maven-plugin:0.31.0:start (docker-start) on project quarkus-integration-test-keycloak-authorization: I/O Error: Unable to pull 'quay.io/keycloak/keycloak:12.0.4' from registry 'quay.io' : filesystem layer verification failed for digest sha256:76cc9a281497e32fd06bee0e3bbcb1d9610db913f8811425beb8f491af8af0ac -> [Help 1]
```
will retry tomorrow

File: extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientImpl.java [hunk 1]
```diff
@@ -93,7 +93,8 @@ public OidcClientImpl(WebClient client, String tokenRequestUri, String grantType
                 Uni<HttpResponse<Buffer>> response = request.sendBuffer(OidcCommonUtils.encodeForm(body))
                         .onFailure(ConnectException.class)
                         .retry()
-                        .atMost(3);
+                        .atMost(3)
```
Comment (pedroigor) [inline extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientImpl.java:96]: Would make sense to make this configurable?

File: extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientRecorder.java [hunk 2]
```diff
@@ -110,34 +109,32 @@ protected static OidcClient createOidcClient(OidcClientConfig oidcConfig, String
         try {
             OidcCommonUtils.verifyCommonConfiguration(oidcConfig, false);
         } catch (Throwable t) {
+            LOG.debug(t.getMessage());
```
Comment (pedroigor) [inline extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientRecorder.java:112]: Would make more sense to log the stack trace?

File: extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientImpl.java (hunk 1 above)
Comment (sberyozkin) [reply extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientImpl.java:96]: @pedroigor Hi Pedro - we already have it indirectly configurable when OIDC connection is established at the start up (`connection-delay` property - it is divided by 2, ex, if it is `20s` then 10 attempts to connect will be made every 2 seconds).
This case is different - there was a test failing (and I've only seen it once) where the connection to KC was dropped - so here it is a try to recover for a very short period of time - this is when a client is awaiting for response, so I was not sure it should be configurable. But I don't really mind - numbers do not look that great :-), I'll have a look

File: extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientRecorder.java (hunk 2 above)
Comment (sberyozkin) [reply extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientRecorder.java:112]: @pedroigor, probably not - the message would say something like `auth-server-url must be set, etc` so the stacktrace would just show the flow via Oidc(Client)Recorder to `OidcCommonUtils` - I believe the stacktraces are not logged in general in Quarkus

**Review by pedroigor — APPROVED** [review]

**Comment by sberyozkin** [comment]
Hi @pedroigor oh thanks :-), was just about to update you I've added a config property :-)
cheers

**Comment by sberyozkin** [comment]
Adding a backport label given that before 1.13.1 it was failing immediately in case of the invalid/badly structured URL which this PR restores - but it is probably not critical

**Pull request merged**